k256/src/arithmetic/affine_op.rs

```diff
@@ -0,0 +1,423 @@
+        })
+    }
+}
+
+impl Neg for PowdrAffinePoint {
+    type Output = PowdrAffinePoint;
+
+    fn neg(self) -> PowdrAffinePoint {
+        PowdrAffinePoint::neg(&self)
+    }
+}
+
+impl Mul<Scalar> for PowdrAffinePoint {
+    type Output = PowdrAffinePoint;
+
+    fn mul(self, other: Scalar) -> PowdrAffinePoint {
+        mul(&self, &other)
+    }
+}
+
+impl PowdrAffinePoint {
+    /// Ruduces the coordinates of the point to their canonical form.
+    pub fn normalize_coordinates(&self) -> Self {
+        PowdrAffinePoint(AffinePoint {
+            x: self.0.x.normalize(),
+            y: self.0.y.normalize(),
+            infinity: self.0.infinity,
+        })
+    }
+
+    /// Double the point.
+    pub fn double(self) -> PowdrAffinePoint {
+        if self.0.y.normalizes_to_zero().into() {
+            return PowdrAffinePoint(AffinePoint::IDENTITY);
+        }
+
+        let num = FieldElement::from(3u64) * self.0.x.square();
+        let denom = FieldElement::from(2u64) * self.0.y;
+        let lambda = num * denom.invert().unwrap();
+
+        let x3 = lambda.square() - FieldElement::from(2u64) * self.0.x;
+        let y3 = lambda * (self.0.x + x3.negate(3)) - self.0.y;
+
+        PowdrAffinePoint(AffinePoint {
+            x: x3.normalize_weak(),
+            y: y3.normalize_weak(),
+            infinity: 0,
+        })
+    }
+
+    fn neg(&self) -> Self {
+        PowdrAffinePoint(AffinePoint {
+            x: self.0.x,
+            y: self.0.y.negate(1).normalize(),
+            infinity: self.0.infinity,
+        })
+    }
+
+    fn conditional_select(a: &Self, b: &Self, choice: Choice) -> Self {
+        PowdrAffinePoint(AffinePoint {
+            x: FieldElement::conditional_select(&a.0.x, &b.0.x, choice),
+            y: FieldElement::conditional_select(&a.0.y, &b.0.y, choice),
+            infinity: a.0.infinity,
+        })
+    }
+
+    /// Calculates SECP256k1 endomorphism: `self * lambda`.
+    pub fn endomorphism(&self) -> Self {
+        Self(AffinePoint {
+            x: self.0.x * ENDOMORPHISM_BETA,
+            y: self.0.y,
+            infinity: self.0.infinity,
+        })
+    }
+    /// Returns the x-coordinate of the point.
+    pub fn x(&self) -> FieldElement {
+        self.0.x
+    }
+
+    /// Returns the y-coordinate of the point.
+    pub fn y(&self) -> FieldElement {
+        self.0.y
+    }
+}
+
+#[inline(always)]
+fn mul(x: &PowdrAffinePoint, k: &Scalar) -> PowdrAffinePoint {
+    lincomb(&[(*x, *k)])
+}
+
+/// multi scalar multiplication using Pippenger's algorithm
+pub fn lincomb<const N: usize>(
+    points_and_scalars: &[(PowdrAffinePoint, Scalar); N],
+) -> PowdrAffinePoint {
+    let mut tables = [(LookupTable::default(), LookupTable::default()); N];
+    let mut digits = [(
+        Radix16Decomposition::<33>::default(),
+        Radix16Decomposition::<33>::default(),
+    ); N];
+
+    lincomb_pippenger(points_and_scalars, &mut tables, &mut digits)
+}
+
+fn lincomb_pippenger(
```

Comment:
This appears to be duplicating code from `k256/src/arithmetic/mul.rs`. I need to take a closer look, but is this algorithm equally applicable for affine points? If yes, would it be possible to make `mul.rs` generic? I think for most operations `conditional_select`, there are already traits, but I think there are exceptions (e.g. `endomorphism()`).